toolchain/check/handle_class.cpp

```diff
@@ -221,21 +221,27 @@ auto HandleBaseDecl(Context& context, Parse::NodeId parse_node) -> bool {
         context.sem_ir().StringifyType(base_type_id));
   });
 
+  SemIR::NameScopeId base_scope_id = SemIR::NameScopeId::Invalid;
   if (base_type_id != SemIR::TypeId::Error) {
     // For now, we treat all types that aren't introduced by a `class`
     // declaration as being final classes.
     // TODO: Once we have a better idea of which types are considered to be
     // classes, produce a better diagnostic for deriving from a non-class type.
     auto base_class = context.types().TryGetAs<SemIR::ClassType>(base_type_id);
-    if (!base_class ||
-        context.classes().Get(base_class->class_id).inheritance_kind ==
-            SemIR::Class::Final) {
+    auto* base_class_info =
+        base_class ? &context.classes().Get(base_class->class_id) : nullptr;
+    if (!base_class_info ||
+        base_class_info->inheritance_kind == SemIR::Class::Final) {
```

Comment:
This took me a few reads to read correctly. What would you think of something like a `GetBaseScopeId` helper function in order to reduce the flow control paths? Alternately, what about swapping the if/else so that you use base_scope_info next to where you verify it's non-null? I'd prefer the former though, because I think it'd be easier to read -- e.g., I think you could get something like:

```
if (base_type_id == SemIR::TypeId::Error) {
  return SemIR::NameScopeId::Invalid;
}
auto base_class = context.types().TryGetAs<SemIR::ClassType>(base_type_id);
if (base_class) {
  auto base_class_info = context.classes().Get(base_class->class_id);
  if (base_class_info.inheritance_kind == SemIR::Class::Final) {
    CARBON_CHECK(base_class_info->scope_id.is_valid())
        << "Complete class should have a scope";
    return base_class_info.scope_id;
  }
}
// Emit diagnostic
return SemIR::NameScopeId::Invalid;
```

The nullness of base_class_info goes away in that approach.